src/core/collision.cpp

```diff
@@ -611,27 +657,61 @@ void handle_collisions() {
           p1->p.rotation = ROTATION_X | ROTATION_Y | ROTATION_Z;
           p2->p.rotation = ROTATION_X | ROTATION_Y | ROTATION_Z;
 
-          // The vs placement is done by the node on which p1 is non-ghost
-          if (!p1->l.ghost) {
-            bind_at_point_of_collision_calc_vs_pos(p1, p2, pos1, pos2);
+          // Positions of the virtual sites
+          bind_at_point_of_collision_calc_vs_pos(p1, p2, pos1, pos2);
+
+          // place virtual sites on the node where the base particle is not a
+          // ghost
+          if (!p1->l.ghost)
             place_vs_and_relate_to_particle(current_vs_pid, pos1, c.pp1,
                                             initial_pos);
-            current_vs_pid++;
+          else // update the books
+            added_particle(current_vs_pid);
+
+          // Increment counter
+          current_vs_pid++;
+
+          // Same for particle 2
+          if (!p2->l.ghost)
             place_vs_and_relate_to_particle(current_vs_pid, pos2, c.pp2,
                                             initial_pos);
-            current_vs_pid++;
-            bind_at_poc_create_bond_between_vs(current_vs_pid, c);
-          } else { // Just update the books
+          else // update the books
             added_particle(current_vs_pid);
-            current_vs_pid++;
 
-            added_particle(current_vs_pid);
-            current_vs_pid++;
-          }
+          // Increment counter
+          current_vs_pid++;
+
+          // Create bonds between the vs.
+
+          bind_at_poc_create_bond_between_vs(current_vs_pid, c);
         }
+
         if (collision_params.mode & COLLISION_MODE_GLUE_TO_SURF) {
+          // If particles are made inert by a type change on collision
+          if (collision_params.part_type_after_glueing !=
+              collision_params.part_type_to_be_glued) {
+            // We skip the pair if one of the particles has already reacted
+            // but we still increase the particle counters, as other nodes
+            // can not always know whether or not a vs is placed
+            if ((p1->p.type == collision_params.part_type_after_glueing) ||
+                (p2->p.type == collision_params.part_type_after_glueing)) {
+              added_particle(current_vs_pid);
+              current_vs_pid++;
+              continue;
+            }
+          }
+
           Vector3d pos;
-          const int pid = glue_to_surface_calc_vs_pos(p1, p2, pos);
+          const int pid = glue_to_surface_calc_vs_pos(*p1, *p2, pos);
+
+          // Add a bond between the centers of the colliding particles
+          // The bond is placed on the node that has p1
+          if (!local_particles[c.pp1]->l.ghost) {
```

Comment:
Why do we use `local_particles[c.pp1]` here and not simply `p1`?
Also below, there is another occurrence of a deref via `local_particles`. But also a direct use of `p1`.